### PR TITLE
CLM-19069 New index task to save a module descriptor for Nexus IQ.

### PR DESCRIPTION
**Pull request by guillermo-varela**
Brings the "index" feature from the Maven plugin so a XML can be generated and later used by Sonatype CI tools: https://help.sonatype.com/integrations/sonatype-clm-for-maven#SonatypeCLMforMaven-CreatingaComponentIndex

cc @bhamail / @DarthHater / @guillermo-varela / @shaikhu

### CODE REVIEW

File: src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+      if (modules.size() == 1) {
+        moduleIoManager.writeModule(file, modules.get(0));
+        files.add(file);
+      }
+      else {
+        File parent = new File(extension.getModuleFile()).getParentFile();
+        String filename = Files.getNameWithoutExtension(file.getName());
+        for (Module module : modules) {
+          File moduleFile = new File(parent, filename + "-" + module.getId().replace(':', '-') + ".xml");
```
[the remark this answers is not in the document]
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java:71]: ignore

[the remark this answers is not in the document]
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java:71]: @sonatype-lift ignore

File: src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    dependenciesFinder = new DependenciesFinder();
+    moduleIoManager = new ModuleIoManager(log);
+  }
+
+  @TaskAction
+  public void saveModule() {
+    try {
+      List<Module> modules =
+          dependenciesFinder.findModules(getProject(), extension.isAllConfigurations(), extension.getModulesExcluded());
+      File file = new File(extension.getModuleFile());
```
[the remark this answers is not in the document]
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java:60]: @sonatype-lift ignore

File: src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  @TaskAction
+  public void saveModule() {
+    try {
+      List<Module> modules =
+          dependenciesFinder.findModules(getProject(), extension.isAllConfigurations(), extension.getModulesExcluded());
+      List<File> files = new ArrayList<>(modules.size());
+
+      for (Module module : modules) {
+        File file = Paths.get(module.getPathname(), "build", "sonatype-clm", "module.xml").toFile();
```
[the remark this answers is not in the document]
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java:63]: @sonatype-lift ignore

File: src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java [hunk 4]
```diff
@@ -71,13 +73,17 @@
...
+        .flatMap(configuration -> getDependencies(project, configuration,
             resolvedConfiguration -> resolvedConfiguration.getFirstLevelModuleDependencies().stream()))
-        .collect(Collectors.toCollection(LinkedHashSet::new));
+        .collect(Collectors.toMap(ResolvedDependency::getName, Function.identity(), (existing, replacement) -> {
+          if (StringUtils.containsAny(replacement.getConfiguration().toLowerCase(Locale.ROOT), "runtime", "release")) {
+            return replacement;
+          }
+          return existing;
+        }, LinkedHashMap::new)).values());
   }
```
Comment (mr-zepol) [inline src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java:87]: Persona preference, this block of code is huge, so maybe we can split into smaller parts? What about the collectors part, creating a method for it

File: src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        moduleIoManager.writeModule(file, module);
```
Comment (mr-zepol) [inline src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java:64]: it does not let me add a comment in the line above, but Just wanted to say that `sonatype-clm` and `module.xml` are repeated in some places maybe  nice idea would be to have constants for those values

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTask.java [hunk 6]
```diff
@@ -70,8 +72,10 @@ public void audit() {
     boolean hasVulnerabilities;
 
     try (OssindexClient ossIndexClient = buildOssIndexClient()) {
-      Set<ResolvedDependency> dependencies =
-          dependenciesFinder.findResolvedDependencies(getProject(), extension.isAllConfigurations());
+      Set<ResolvedDependency> dependencies = getProject().getAllprojects().stream()
+          .flatMap(
+              project -> dependenciesFinder.findResolvedDependencies(project, extension.isAllConfigurations()).stream())
+          .collect(Collectors.toCollection(LinkedHashSet::new));
```
Comment (mr-zepol) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTask.java:78]:
```suggestion
      Set<ResolvedDependency> dependencies = getProject().getAllprojects().stream()
          .flatMap(
              project -> dependenciesFinder.findResolvedDependencies(project, extension.isAllConfigurations())
              .stream())
          .collect(Collectors.toCollection(LinkedHashSet::new));
```

Personal Preference -> I feel this way the code can be easier to read

File: src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java (hunk 4 above)
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java]: > Persona preference, this block of code is huge, so maybe we can split into smaller parts? What about the collectors part, creating a method for it

Done: 6728e6f38ec379945b7790d2f4e2cd2e591c13f9

File: src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java (hunk 5 above)
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java:68]: > it does not let me add a comment in the line above, but Just wanted to say that `sonatype-clm` and `module.xml` are repeated in some places maybe nice idea would be to have constants for those values

Done: 6728e6f38ec379945b7790d2f4e2cd2e591c13f9, 2b3e9da1aa23a457fb8d4d8b3ff60e592568a631

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTask.java (hunk 6 above)
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTask.java:78]: > Personal Preference -> I feel this way the code can be easier to read

I actually find it easier to read the way it currently is.

File: README.md [hunk 7]
```diff
@@ -56,7 +56,7 @@ Gradle can be used to build projects developed in various programming languages.
 - Edit its `build.gradle` file adding this:
 ```
 plugins {
-  id 'org.sonatype.gradle.plugins.scan' version '2.0.10' // Update the version as needed
+  id 'org.sonatype.gradle.plugins.scan' version '2.1.1' // Update the version as needed
```
Comment (mealingr) [inline README.md:59]: We're updating the minor version because we're adding a backwards compatible feature right? Should the patch version be reset to 0? (or was there also some bug fixes?)

File: README.md [hunk 8]
```diff
@@ -116,6 +116,14 @@ nexusIQScan {
 - Open Terminal on the project's root and run `./gradlew nexusIQScan`
 - You should see the scan report URL report on Terminal.
 
+### Nexus IQ Index
+Allow to save information about the dependencies of a project into a module information file (module.xml) that Sonatype CI tools can use to include these dependencies in a scan.
```
Comment (mealingr) [inline README.md:120]:
```suggestion
Allows you to save information about the dependencies of a project into a module information file (module.xml) that Sonatype CI tools can use to include these dependencies in a scan.
```

File: README.md [hunk 9]
```diff
@@ -116,6 +116,14 @@ nexusIQScan {
...
+```
+nexusIQIndex {
+     modulesExcluded = ['module-1', 'module-2'] // Optional. For multi-module projects, the names of the sub-modules to exclude from indexing.
```
Comment (mealingr) [inline README.md:123]: Out of curiosity is this meant to be the equivalent of how maven can exclude modules during a build? e.g.

```
mvn com.sonatype.clm:clm-maven-plugin:index -pl "!clm-maven-plugin-its"
```

will exclude the `clm-maven-plugin-its` module from being indexed

File: src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      for (Module module : modules) {
+        File file = Paths.get(module.getPathname(), "build", SONATYPE_CLM_FOLDER, MODULE_XML_FILE).toFile();
+        moduleIoManager.writeModule(file, module);
+        files.add(file);
+      }
+
+      log.info("Saved module information to {}", StringUtils.join(files, ", "));
+    }
+    catch (Exception e) {
+      throw new GradleException("Could not save the module descriptor for the project: " + e.getMessage(), e);
```
Comment (mealingr) [inline src/main/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTask.java:75]: To match the message above
```suggestion
      throw new GradleException("Could not save the module information for the project: " + e.getMessage(), e);
```

File: src/main/java/org/sonatype/gradle/plugins/scan/ScanPlugin.java [hunk 11]
```diff
@@ -29,11 +31,17 @@
...
+    project.getExtensions().create("nexusIQScan", NexusIqPluginScanExtension.class, project);
     NexusIqScanTask nexusIqScanTask = project.getTasks().create("nexusIQScan", NexusIqScanTask.class);
     nexusIqScanTask.setGroup(SONATYPE_GROUP);
     nexusIqScanTask.setDescription("Scan and evaluate the dependencies of the project using Nexus IQ Server.");
 
+    project.getExtensions().create("nexusIQIndex", NexusIqPluginIndexExtension.class, project);
+    NexusIqIndexTask nexusIqIndexTask = project.getTasks().create("nexusIQIndex", NexusIqIndexTask.class);
+    nexusIqIndexTask.setGroup(SONATYPE_GROUP);
+    nexusIqIndexTask.setDescription("Saves information about the dependencies of a project into a module information "
+        + "file that Sonatype CI tools can use to include these dependencies in a scan.");
```
Comment (mealingr) [inline src/main/java/org/sonatype/gradle/plugins/scan/ScanPlugin.java:43]: since there might be multiple
```suggestion
    nexusIqIndexTask.setDescription("Saves information about the dependencies of a project into module information "
        + "(module.xml) files that Sonatype CI tools can use to include these dependencies in a scan.");
```

File: src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java [hunk 12]
```diff
@@ -71,13 +74,12 @@
 
   private static final String ATTRIBUTES_SUPPORTED_GRADLE_VERSION = "4.0";
 
-  public Set<ResolvedDependency> findResolvedDependencies(Project rootProject, boolean allConfigurations) {
-    return new LinkedHashSet<>(rootProject.getAllprojects()).stream()
-        .flatMap(project -> new LinkedHashSet<>(project.getConfigurations()).stream())
+  public Set<ResolvedDependency> findResolvedDependencies(Project project, boolean allConfigurations) {
+    return new LinkedHashSet<>(new LinkedHashSet<>(project.getConfigurations()).stream()
```
Comment (mealingr) [inline src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java:78]: Why did this change out of curiosity? If I understand correct, previously it was considering the configurations of all projects from `getAllprojects`, and now we're just looking at configurations from the root project?

Comment (mealingr) [reply src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java:78]: > Why did this change out of curiosity?

Ah nvm I think I understand, looking at the non-test files where it's called
- This is already looping through `rootProject.allprojects` https://github.com/sonatype-nexus-community/scan-gradle-plugin/blob/6728e6f38ec379945b7790d2f4e2cd2e591c13f9/src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java#L96-L113
- You changed this to loop through all `getProject().getAllprojects()` https://github.com/sonatype-nexus-community/scan-gradle-plugin/blob/f2b3cd62add9d7c8fb87e644060e5280f86f4a83/src/main/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTask.java#L75-L78

👍

File: src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java [hunk 13]
```diff
@@ -119,7 +121,11 @@
 
   @VisibleForTesting
   Module buildModule(Project project) {
-    Module module = new Module().setIdKind("gradle").setPathname(project.getProjectDir());
+    Module module = new Module()
+        .setIdKind("gradle")
+        .setPathname(project.getProjectDir())
+        .setBuilderInfo(Module.BI_CLM_TOOL, "gradle")
+        .setBuilderInfo(Module.BI_CLM_VERSION, PluginVersionUtils.getPluginVersion());
```
Comment (mealingr) [inline src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java:128]: Why is this needed out of curiosity?

File: src/test/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTaskTest.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+    NexusIqIndexTask task = buildIndexTask();
+    task.setDependenciesFinder(dependenciesFinderMock);
+    task.setModuleIoManager(moduleIoManagerMock);
+    task.saveModule();
+
+    verify(dependenciesFinderMock).findModules(any(Project.class), eq(false), anySet());
+    verify(moduleIoManagerMock).writeModule(file1, module1);
+    verify(moduleIoManagerMock).writeModule(file2, module2);
+  }
```
Comment (mealingr) [inline src/test/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTaskTest.java:98]: Could it be worth adding a test which has some modules excluded as well? I mean I see `DependenciesFinder.findModules` which uses the excludes already existed, but it could be more to test that your configuration `NexusIqPluginIndexExtension` accepts them and passes them to that method in `NexusIqIndexTask.saveModule`

File: README.md (hunk 9 above)
Comment (guillermo-varela) [reply README.md:123]: > Out of curiosity is this meant to be the equivalent of how maven can exclude modules during a build? e.g.
> 
> ```
> mvn com.sonatype.clm:clm-maven-plugin:index -pl "!clm-maven-plugin-its"
> ```
> 
> will exclude the `clm-maven-plugin-its` module from being indexed

Indeed. I'm using `modulesExcluded` because it's a name already used in another existing task.

File: src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java (hunk 13 above)
Comment (guillermo-varela) [reply src/main/java/org/sonatype/gradle/plugins/scan/common/DependenciesFinder.java:128]: > Why is this needed out of curiosity?

Not sure if needed, but the Maven plugin is adding those fields in the `module.xml` files: https://github.com/sonatype/clm-maven-plugin/blob/13783077fe95d411ed641eb53dfb7a6782838b3b/clm-maven-plugin/src/main/java/com/sonatype/insight/maven/IndexMojo.java#L45-L46

File: src/test/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTaskTest.java (hunk 14 above)
Comment (guillermo-varela) [reply src/test/java/org/sonatype/gradle/plugins/scan/nexus/iq/index/NexusIqIndexTaskTest.java:100]: > Could it be worth adding a test which has some modules excluded as well? I mean I see `DependenciesFinder.findModules` which uses the excludes already existed, but it could be more to test that your configuration `NexusIqPluginIndexExtension` accepts them and passes them to that method in `NexusIqIndexTask.saveModule`

Done: 77ba696b8fe11a816e4bed62bdcd1b75495e821b, 34c00d70765afbb3da3aa00f1cd6c6799933c2f1

**Review by mealingr — APPROVED** [review]
+1

**Review by mr-zepol — APPROVED** [review]
+1 LGTM
